### PR TITLE
WIP for incremental discrete

### PR DESCRIPTION
**Pull request by ProfFan**
WIP for incremental discrete elimination.

### CODE REVIEW

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (ProfFan) [inline gtsam/hybrid/tests/testHybridFactorGraph.cpp:393]: End of 3 step, now we do the incremental

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (ProfFan) [inline gtsam/hybrid/tests/testHybridFactorGraph.cpp:407]: Old discrete result is pushed

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (ProfFan) [inline gtsam/hybrid/tests/testHybridFactorGraph.cpp:433]: DTF4 and the product differs!

**Comment by ProfFan** [comment]
@varunagrawal

File: gtsam/hybrid/tests/testDCGaussianMixtureFactor.cpp
Comment (dellaert) [inline gtsam/hybrid/tests/testDCGaussianMixtureFactor.cpp:102]: Does not make sense to me

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (dellaert) [inline gtsam/hybrid/tests/testHybridFactorGraph.cpp:455]: Where is this defined???

File: gtsam/hybrid/tests/testDCGaussianMixtureFactor.cpp
Comment (ProfFan) [reply gtsam/hybrid/tests/testDCGaussianMixtureFactor.cpp:102]: GaussianConditional here is empty (invalid)

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (ProfFan) [reply gtsam/hybrid/tests/testHybridFactorGraph.cpp]: Sorry in half way of a change

**Comment by dellaert** [comment]
You ask for a review, and you say, it's not correct, but could you maybe comment here what the outputs are, what you are expecting, and why you think it's wrong? Also, CI fails...

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (dellaert) [inline gtsam/hybrid/tests/testHybridFactorGraph.cpp:582]: I think you should do the eliminatePartial, and grab the actual factor, and then compare.

Comment (ProfFan) [reply gtsam/hybrid/tests/testHybridFactorGraph.cpp:603]: I fixed the unit test by comparing the final elimination results.

File: gtsam/base/utilities.cpp
Comment (ProfFan) [inline gtsam/base/utilities.cpp:13]: fix this newline when CI is not busy

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (dellaert) [inline gtsam/hybrid/tests/testHybridFactorGraph.cpp:586]: I would not call these `actual` here but `expected`

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (dellaert) [inline gtsam/hybrid/tests/testHybridFactorGraph.cpp:506]: OK, this should be done now :-)

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (dellaert) [inline gtsam/hybrid/tests/testHybridFactorGraph.cpp:521]: I think we need the graph0 stage now as we discussed. It threw an error, probably because we made a wrong assumption somewhere?

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (dellaert) [inline gtsam/hybrid/tests/testHybridFactorGraph.cpp:435]: kill this obsolete test now?

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (dellaert) [inline gtsam/hybrid/tests/testHybridFactorGraph.cpp:493]: After comments addressed, move to new file?

**Review by dellaert — COMMENTED** [review]
Awesome. 
- There is still a number of things left undone, which should be taken care of. 
- I propose to merge after the comments have been addressed, but in the next PR we should include the discrete factors into the update. I did not want to do this before to focus on one thing at a time.
- please merge in develop so we don't diverge too much
I'll approve after one more iteration.

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (ProfFan) [reply gtsam/hybrid/tests/testHybridFactorGraph.cpp:547]: Because I need a copy so I can make sure that this is not overwritten.

File: gtsam/hybrid/DCGaussianMixtureFactor.cpp
Comment (ProfFan) [reply gtsam/hybrid/DCGaussianMixtureFactor.cpp:72]: got rid of

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (ProfFan) [reply gtsam/hybrid/tests/testHybridFactorGraph.cpp:576]: done

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (ProfFan) [reply gtsam/hybrid/tests/testHybridFactorGraph.cpp:606]: done

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (ProfFan) [reply gtsam/hybrid/tests/testHybridFactorGraph.cpp:586]: done

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (ProfFan) [reply gtsam/hybrid/tests/testHybridFactorGraph.cpp:506]: done

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (ProfFan) [reply gtsam/hybrid/tests/testHybridFactorGraph.cpp:521]: Yes, maybe after merging this?

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (ProfFan) [reply gtsam/hybrid/tests/testHybridFactorGraph.cpp:435]: killed

File: gtsam/hybrid/IncrementalHybrid.cpp
Comment (dellaert) [inline gtsam/hybrid/IncrementalHybrid.cpp:24]: comment on what is happening here

File: gtsam/hybrid/IncrementalHybrid.h
Comment (dellaert) [inline gtsam/hybrid/IncrementalHybrid.h:40]: missing newline

File: gtsam/hybrid/IncrementalHybrid.h
Comment (dellaert) [inline gtsam/hybrid/IncrementalHybrid.h:35]: param ordering.
Also, tell us exactly what will happen in this function, and where the result sits.

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (dellaert) [inline gtsam/hybrid/tests/testHybridFactorGraph.cpp:438]: TEST

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp
Comment (dellaert) [inline gtsam/hybrid/tests/testHybridFactorGraph.cpp:529]: what is "this". Comment on what you are checking below.

**Review by dellaert — APPROVED** [review]
Cool, mostly comments missing. Now is the time to add them. Typing is never the bottleneck

**Review by dellaert — APPROVED** [review]
Cool! Two things:
- Please merge develop into base branch as well so this PR is only the diff
- Resolve the conflict.
I'll approve but please do these things before merging.

**Comment by ProfFan** [comment]
Appears that the macOS CI is broken and I don't know why...

Ping @varunagrawal for help

**Comment by varunagrawal** [comment]
I know why! I'll put in a fix in a bit

File: gtsam/base/utilities.cpp
Comment (varunagrawal) [inline gtsam/base/utilities.cpp:13]: FYI if you run the formatter on this file, it should add the `// namespace gtsam` comment at this point.

File: gtsam/hybrid/HybridFactorGraph.cpp
Comment (varunagrawal) [inline gtsam/hybrid/HybridFactorGraph.cpp:73]: Then we should make Factor::equals virtual. :)

File: gtsam/hybrid/GaussianMixture.cpp
Comment (varunagrawal) [inline gtsam/hybrid/GaussianMixture.cpp:49]: Do we need the regex? I don't see why we can't do something much simpler and future-proof.

File: gtsam/hybrid/DCGaussianMixtureFactor.cpp
Comment (varunagrawal) [inline gtsam/hybrid/DCGaussianMixtureFactor.cpp:57]: Same question about regex here.

File: gtsam/hybrid/HybridFactorGraph.cpp
Comment (ProfFan) [reply gtsam/hybrid/HybridFactorGraph.cpp:73]: Yes, but I don't want to touch it now, thus this comment :)

File: gtsam/hybrid/GaussianMixture.cpp
Comment (ProfFan) [reply gtsam/hybrid/GaussianMixture.cpp:49]: This is very simple and future proof (C++11 -> 20), just a bit slow... But I think this is not a problem for printing, so.

**Comment by ProfFan** [comment]
Thanks @varunagrawal !

File: gtsam/hybrid/GaussianMixture.cpp
Comment (varunagrawal) [reply gtsam/hybrid/GaussianMixture.cpp:49]: But isn't this just reading the print output and then reprinting it with extra indentation? In that case we could simply use the RedirectCout to capture the result of `p->print()` and just return that. That's what I've been doing so far and it works great in terms of readability.

Comment (varunagrawal) [reply gtsam/hybrid/GaussianMixture.cpp:49]: Plus it's a lot simpler and less error prone by someone who's not as good at C++ as you :)

File: gtsam/hybrid/HybridFactorGraph.cpp
Comment (varunagrawal) [reply gtsam/hybrid/HybridFactorGraph.cpp:73]: Sounds good!

**Comment by ProfFan** [comment]
I'll merge now, no need to waste more CI cycles.

**Pull request merged**
